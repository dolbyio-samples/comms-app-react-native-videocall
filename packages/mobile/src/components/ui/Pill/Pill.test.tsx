import { defaultTheme as theme } from '@dolbyio/comms-uikit-common';
import React from 'react';
import { Text } from 'react-native';
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from 'react-test-renderer';

import { render, mergeStyles } from '../../../utils/test-utils';

import Pill from './Pill';

const { colors } = theme;

const testID = 'testID';
const text = 'TestPill';
const label = 'TestLabel';
const textElm1TestID = `${testID}-text`;
const textElm2TestID = `${testID}-text-local`;

const commonExpStyle = {
  display: 'flex',
  flexDirection: 'row',
  paddingLeft: 8,
  paddingRight: 8,
  paddingTop: 2,
  paddingBottom: 2,
  justifyContent: 'center',
  alignItems: 'center',
  alignSelf: 'flex-start',
  borderRadius: 4,
  height: 20,
  maxWidth: 140,
};

const commonTextElmExpStyle = {
  letterSpacing: 0.3,
  lineHeight: 16,
  fontWeight: '600',
  fontFamily: 'Roboto',
  textAlign: 'left',
};

describe('Pill component', () => {
  test('Passes TestID', () => {
    const { getByTestId } = render(<Pill text={text} label={label} testID={testID} />);

    const element = getByTestId(testID);
    expect(element).not.toBeNull();

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(2);

    expect(getByTestId(textElm1TestID)).not.toBeNull();
    expect(getByTestId(textElm2TestID)).not.toBeNull();
  });
  test('matches style with size s', () => {
    const size = 's';
    const expStyles = {
      ...commonExpStyle,
      backgroundColor: colors.white,
    };
    const { getByTestId } = render(<Pill text={text} label={label} active size={size} testID={testID} />);
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    expect(styles).toMatchObject(expStyles);

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(2);

    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 10,
      color: colors.purple[400],
    };
    const textElm1Style = mergeStyles(textElmList[0].props.style);
    expect(textElm1Style).toMatchObject(textElmExpStyle);
    expect(textElmList[0].props.children).toBe(text);
    expect(textElmList[0].props.testID).toBe(textElm1TestID);

    const textElm2Style = mergeStyles(textElmList[1].props.style);
    expect(textElm2Style).toMatchObject(textElmExpStyle);
    expect(textElmList[1].props.children).toBe(`(${label})`);
    expect(textElmList[1].props.testID).toBe(textElm2TestID);
  });
  test('matches style with default active/size prop value', () => {
    const expStyles = {
      ...commonExpStyle,
      backgroundColor: colors.whiteAlpha[50],
    };
    const { getByTestId } = render(<Pill text={text} label={label} testID={testID} />);
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    expect(styles).toMatchObject(expStyles);

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(2);

    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 12,
      color: colors.white,
    };
    const textElm1Style = mergeStyles(textElmList[0].props.style);
    expect(textElm1Style).toMatchObject(textElmExpStyle);
    expect(textElmList[0].props.children).toBe(text);
    expect(textElmList[0].props.testID).toBe(`${testID}-text`);

    const textElm2Style = mergeStyles(textElmList[1].props.style);
    expect(textElm2Style).toMatchObject(textElmExpStyle);
    expect(textElmList[1].props.children).toBe(`(${label})`);
    expect(textElmList[1].props.testID).toBe(`${testID}-text-local`);
  });
  test('matches style with size m and active false', () => {
    const size = 'm';
    const expStyles = {
      ...commonExpStyle,
      backgroundColor: colors.whiteAlpha[50],
    };
    const { getByTestId } = render(<Pill text={text} label={label} size={size} active={false} testID={testID} />);
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    expect(styles).toMatchObject(expStyles);

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(2);

    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 12,
      color: colors.white,
    };
    const textElm1Style = mergeStyles(textElmList[0].props.style);
    expect(textElm1Style).toMatchObject(textElmExpStyle);
    expect(textElmList[0].props.children).toBe(text);
    expect(textElmList[0].props.testID).toBe(`${testID}-text`);

    const textElm2Style = mergeStyles(textElmList[1].props.style);
    expect(textElm2Style).toMatchObject(textElmExpStyle);
    expect(textElmList[1].props.children).toBe(`(${label})`);
    expect(textElmList[1].props.testID).toBe(`${testID}-text-local`);
  });
  test('matches style without label prop', () => {
    const size = 'm';
    const expStyles = {
      ...commonExpStyle,
      backgroundColor: colors.whiteAlpha[50],
    };
    const { getByTestId } = render(<Pill text={text} size={size} active={false} testID={testID} />);
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    expect(styles).toMatchObject(expStyles);

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(1);

    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 12,
      color: colors.white,
    };
    const textElm1Style = mergeStyles(textElmList[0].props.style);
    expect(textElm1Style).toMatchObject(textElmExpStyle);
    expect(textElmList[0].props.children).toBe(text);
    expect(textElmList[0].props.testID).toBe(`${testID}-text`);
  });
  test('matches style without text but label prop', () => {
    const size = 'm';
    const expStyles = {
      ...commonExpStyle,
      backgroundColor: colors.white,
    };
    const { getByTestId } = render(<Pill label={label} size={size} active testID={testID} />);
    const element = getByTestId(testID);
    const styles = mergeStyles(element.props.style);
    expect(styles).toMatchObject(expStyles);

    const textElmList = element.findAllByType(Text);
    expect(textElmList).toHaveLength(2);

    const textElmExpStyle = {
      ...commonTextElmExpStyle,
      fontSize: 12,
      color: colors.purple[400],
    };
    const textElm1Style = mergeStyles(textElmList[0].props.style);
    expect(textElm1Style).toMatchObject(textElmExpStyle);
    expect(textElmList[0].props.children).toBe(undefined);
    expect(textElmList[0].props.testID).toBe(`${testID}-text`);

    const textElm2Style = mergeStyles(textElmList[1].props.style);
    expect(textElm2Style).toMatchObject(textElmExpStyle);
    expect(textElmList[1].props.children).toBe(`(${label})`);
    expect(textElmList[1].props.testID).toBe(`${testID}-text-local`);
  });
});

describe('Pill component snapshot', () => {
  test('should match default param configuration', () => {
    const tree = renderer.create(<Pill testID={testID} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('should match when size/active prop is s/true', () => {
    const size = 's';
    const tree = renderer.create(<Pill text={text} active size={size} testID={testID} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('should match when size/active/label prop is m/false/value', () => {
    const size = 'm';
    const tree = renderer
      .create(<Pill text={text} label={label} active={false} size={size} testID={testID} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
