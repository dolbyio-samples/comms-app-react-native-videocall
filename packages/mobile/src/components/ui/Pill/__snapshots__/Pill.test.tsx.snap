// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pill component snapshot should match default param configuration 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "borderRadius": 4,
        "display": "flex",
        "flexDirection": "row",
        "height": 20,
        "justifyContent": "center",
        "maxWidth": 140,
        "paddingBottom": 2,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 2,
      },
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0.04)",
      },
    ]
  }
  testID="testID"
>
  <View
    style={
      Object {
        "flexShrink": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "letterSpacing": 0.3,
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
      testID="testID-text"
    />
  </View>
</View>
`;

exports[`Pill component snapshot should match when size/active prop is s/true 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "borderRadius": 4,
        "display": "flex",
        "flexDirection": "row",
        "height": 20,
        "justifyContent": "center",
        "maxWidth": 140,
        "paddingBottom": 2,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 2,
      },
      Object {
        "backgroundColor": "#FFFFFF",
      },
    ]
  }
  testID="testID"
>
  <View
    style={
      Object {
        "flexShrink": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 10,
            "fontWeight": "600",
            "letterSpacing": 0.3,
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#AA33FF",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
      testID="testID-text"
    >
      TestPill
    </Text>
  </View>
</View>
`;

exports[`Pill component snapshot should match when size/active/label prop is m/false/value 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "alignSelf": "flex-start",
        "borderRadius": 4,
        "display": "flex",
        "flexDirection": "row",
        "height": 20,
        "justifyContent": "center",
        "maxWidth": 140,
        "paddingBottom": 2,
        "paddingLeft": 8,
        "paddingRight": 8,
        "paddingTop": 2,
      },
      Object {
        "backgroundColor": "rgba(255, 255, 255, 0.04)",
      },
    ]
  }
  testID="testID"
>
  <View
    style={
      Object {
        "flexShrink": 1,
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "letterSpacing": 1.22,
          },
          Object {
            "fontSize": 12,
            "fontWeight": "600",
            "letterSpacing": 0.3,
            "lineHeight": 16,
          },
          false,
          Object {
            "color": "#FFFFFF",
            "fontFamily": "Roboto",
            "textAlign": "left",
          },
        ]
      }
      testID="testID-text"
    >
      TestPill
    </Text>
  </View>
  <Text
    style={
      Array [
        Object {
          "letterSpacing": 1.22,
        },
        Object {
          "fontSize": 12,
          "fontWeight": "600",
          "letterSpacing": 0.3,
          "lineHeight": 16,
        },
        false,
        Object {
          "color": "#FFFFFF",
          "fontFamily": "Roboto",
          "textAlign": "left",
        },
      ]
    }
    testID="testID-text-local"
  >
    (TestLabel)
  </Text>
</View>
`;
